packages/js/wasm/src/WasmWrapper.ts

```diff
@@ -172,12 +181,21 @@ export class WasmWrapper implements Wrapper {
         env: options.env ? msgpackEncode(options.env) : EMPTY_ENCODED_OBJECT,
       };
 
-      const abort = (message: string) => {
-        throw Error(
-          `WasmWrapper: Wasm module aborted execution.\nURI: ${options.uri.uri}\n` +
-            `Method: ${method}\n` +
-            `Args: ${JSON.stringify(args, null, 2)}\nMessage: ${message}.\n`
-        );
+      const abort = (
+        message: string,
+        code: WrapErrorCode = WrapErrorCode.WRAPPER_INVOKE_ABORTED,
+        source?: ErrorSource
+      ) => {
+        const prev = WrapError.parse(message);
```

Comment:
Parsing message is central to this PR and it has little to do with error codes.

Let's review the status quo:

```ts
    const message = `Error: WasmWeb3Api: Wasm module aborted execution.
    URI: w3://ipfs/QmRZrQkmjK9TnC8fHwJCf7YRqMFiyPvcN5vcYc9ZSXQ9tB
    Module: query
    Method: queryEnv
    Input: {
      "arg": "string"
    }
    Message: __w3_abort: 129
    File: ~lib/@web3api/wasm-as/msgpack/ReadDecoder.ts
    Location: [166,5].`
    
    const abort = (message: string) => {
        throw Error(
          `WasmWrapper: Wasm module aborted execution.\nURI: ${options.uri.uri}\n` +
            `Method: ${method}\n` +
            `Args: ${JSON.stringify(args, null, 2)}\nMessage: ${message}.\n`
        );
      };
      
      abort(message);
```

Any time an exception is thrown during a sub-invocation, `message` is a stringified error. That means errors get nested within errors, leading to output that looks something like this:

```ts
    Error: WasmWeb3Api: Wasm module aborted execution.
    URI: w3://ipfs/QmRZrQkmjK9TnC8fHwJCf7YRqMFiyPvcN5vcYc9ZSXQ9tB
    Module: query
    Method: queryEnv
    Input: {
      "arg": "string"
    }
    Message: Error: WasmWeb3Api: Wasm module aborted execution.
    URI: w3://ipfs/QmRZrQkmjK9TnC8fHwJCf7YRqMFiyPvcN5vcYc9ZSXQ9tB
    Module: query
    Method: queryEnv
    Input: {
      "arg": "string"
    }
    Message: __w3_abort: 129
    File: ~lib/@web3api/wasm-as/msgpack/ReadDecoder.ts
    Location: [166,5].
    File: ~lib/@web3api/wasm-as/msgpack/ReadDecoder.ts
    Location: [166,5].
```

Now imagine that happening multiple layers deep (e.g., a sub-sub-sub-invocation)!

And recall that message can be any message or any stringified polywrap error (either a string as shown above in the status quote, or in this PR a `WrapError`).

This is why parsing the message is necessary.

At first glance, you might wonder why we don't just capture the error after the sub-invocation within `__wrap_subinvoke`. The answer is simply that by this time the error messages are already nested. `__wrap_abort` is called during an invocation, not after it.

The best way around parsing is to allow errors to cross the Wasm boundary so we are working with msgpack encoded bytes instead of strings. I think that's definitely something we should look at in the future, but I think parsing is a nice incremental improvement over the status quo.

Reply:
If you were instead asking why we parse for error codes at all in the `parse` regex, it is because the error code could be any wrap error code. There is no reason to expect a parsed error code to be  `WRAPPER_INVOKE_ABORTED`. 

To clarify, an error code will be present if and only if the error was thrown during a subinvocation and then not handled by the higher level invocation (e.g. re-thrown). In that case, `message` could be a stringified version of any `WrapError` that can be returned from `client.invoke`.

Reply:
No, I'm saying that the `abort` function is only called in `imports.ts`.
It is called in many places there where the message is just a simple string, and the only time it is not is when it is called from within `__wrap_abort` where the code is WRAPPER_INVOKE_ABORTED.

So you could rewrite it like this:
```typescript
const prev = WrapError.parse(message);
```
Into this:
```typescript
const prev = code === WrapErrorCode.WRAPPER_INVOKE_ABORTED
  ? WrapError.parse(message)
  : new WrapError(message)
```
Or something to that effect.

Since the only other code that gets passed here is WRAPPER_INTERNAL_ERROR, you could also write it like this.
```typescript
const prev = code === WrapErrorCode.WRAPPER_INTERNAL_ERROR
  ? new WrapError(message)
  : WrapError.parse(message)
```

Ofc, it is a bit of a bad practice to rely on enums like this, so a better solution could be to replace `abort` with `abortWithInvokeAborted` and `abortWithInternalError`, e.g.
```
  const abortWithInvokeAborted = (
        serializedMessage: string,
        source: ErrorSource
      ) => {
        const code = WrapErrorCode.WRAPPER_INVOKE_ABORTED;
        const prev = WrapError.parse(message);
        ....
    };
    
   const abortWithInternalError = (
        message: string
      ) => {
        const code = WrapErrorCode.WRAPPER_INTERNAL_ERROR;
        const prev = new WrapError(message);
        ....
    }
```

Though I don't think you even need prev for the internal error case.

Reply:
Ohhhh okay. That makes complete sense. I totally misunderstood your initial question. I replaced `abort` with two methods: `abortWithInvokeAborted` and `abortWithInternalError`.